File: sdks/ts/memory/src/http/sse.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * SSE helper for the memory daemon. Returns a {@link Response} whose
 * body is a `ReadableStream` the handler can push frames onto via the
 * returned writer. The writer is safe to call from async code; closing
 * the abort signal released by the runtime drains the stream cleanly.
 */

export type SseWriter = {
  sendRaw: (event: string, data: string) => boolean
  sendJson: (event: string, payload: unknown) => boolean
  close: () => void
  readonly closed: boolean
}

export type SseSession = {
  readonly response: Response
  readonly writer: SseWriter
  /** Resolves when the client disconnects or the writer closes. */
  readonly done: Promise<void>
}

const encoder = new TextEncoder()

const formatFrame = (event: string, data: string): string => {
  const lines: string[] = []
  lines.push(`event: ${event}`)
  for (const line of data.split('\n')) {
    lines.push(`data: ${line}`)
  }
  lines.push('', '')
  return lines.join('\n')
}

/**
 * Start an SSE stream. The caller receives a Response pre-configured
 * with the correct headers plus a writer that pushes frames.
 *
 * @param signal Abort signal propagated by the runtime when the client
 *   disconnects. Closing the signal also closes the writer.
 */
export const startSse = (signal: AbortSignal | undefined): SseSession => {
  let controllerRef: ReadableStreamDefaultController<Uint8Array> | undefined
  let closed = false
  let resolveDone: (() => void) | undefined
  const done = new Promise<void>((resolve) => {
    resolveDone = resolve
  })

  const stream = new ReadableStream<Uint8Array>({
    start(controller) {
      controllerRef = controller
    },
    cancel() {
      closed = true
      resolveDone?.()
    },
  })

  const send = (event: string, data: string): boolean => {
    if (closed) return false
    try {
      controllerRef?.enqueue(encoder.encode(formatFrame(event, data)))
      return true
    } catch {
      closed = true
      resolveDone?.()
      return false
    }
  }

  const onAbort = (): void => {
    if (closed) return
    closed = true
    try {
      controllerRef?.close()
    } catch {
      /* ignore */
    }
    resolveDone?.()
  }
  if (signal !== undefined) {
    if (signal.aborted) onAbort()
    else signal.addEventListener('abort', onAbort, { once: true })
  }

  const writer: SseWriter = {
    sendRaw: (event, data) => send(event, data),
    sendJson: (event, payload) => send(event, JSON.stringify(payload)),
    close: () => onAbort(),
    get closed() {
      return closed
    },
  }

  const response = new Response(stream, {
    status: 200,
    headers: {
      'content-type': 'text/event-stream',
      'cache-control': 'no-cache',
      connection: 'keep-alive',
      'x-accel-buffering': 'no',
    },
  })

  return { response, writer, done }
}

File: go/internal/httpd/sse.go
// SPDX-License-Identifier: Apache-2.0

package httpd

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

// SSEWriter is a small wrapper that emits text/event-stream frames
// and flushes after every write. Callers use [SSEStart] to set the
// canonical headers and obtain a writer.
type SSEWriter struct {
	w  http.ResponseWriter
	fl http.Flusher
}

// SSEStart sets the SSE response headers and returns a writer ready
// for [SSEWriter.SendJSON] / [SSEWriter.SendRaw]. Returns nil and
// writes a 500 Problem+JSON when the response writer cannot stream.
func SSEStart(w http.ResponseWriter) *SSEWriter {
	flusher, ok := w.(http.Flusher)
	if !ok {
		InternalError(w, "streaming not supported")
		return nil
	}
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-store")
	w.Header().Set("X-Accel-Buffering", "no")
	w.Header().Set("Connection", "keep-alive")
	w.WriteHeader(http.StatusOK)
	flusher.Flush()
	return &SSEWriter{w: w, fl: flusher}
}

// SendRaw emits an SSE frame with the given event and pre-encoded
// data. Returns the underlying error so callers can detect client
// disconnect.
func (s *SSEWriter) SendRaw(event, data string) error {
	if s == nil {
		return io.ErrClosedPipe
	}
	if _, err := fmt.Fprintf(s.w, "event: %s\ndata: %s\n\n", event, data); err != nil {
		return err
	}
	s.fl.Flush()
	return nil
}

// SendJSON marshals payload and emits an SSE frame.
func (s *SSEWriter) SendJSON(event string, payload any) error {
	if s == nil {
		return io.ErrClosedPipe
	}
	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	return s.SendRaw(event, string(data))
}
